src/overlays/actors/ovl_En_Ossan/z_en_ossan.c

```diff
@@ -37,227 +114,2413 @@ const ActorInit En_Ossan_InitVars = {
+}
+
+void EnOssan_MainActionFunc(EnOssan* this, GlobalContext* globalCtx) {
+    static EnOssanStateFunc stateFunc[] = {
+        EnOssan_State_Idle,
+        EnOssan_State_StartConversation,
+        EnOssan_State_FacingShopkeeper,
+        EnOssan_State_TalkingToShopkeeper,
+        EnOssan_State_LookToLeftShelf,
+        EnOssan_State_LookToRightShelf,
+        EnOssan_State_BrowseLeftShelf,
+        EnOssan_State_BrowseRightShelf,
+        EnOssan_State_LookFromShelfToShopkeeper,
+        EnOssan_State_ItemSelected,
+        EnOssan_State_SelectMilkBottle,
+        EnOssan_State_SelectWeirdEgg,
+        EnOssan_State_SelectUnimplementedItem,
+        EnOssan_State_SelectBombs,
+        EnOssan_State_CantGetItem,
+        EnOssan_State_GiveItemWithFanfare,
+        EnOssan_State_ItemPurchased,
+        EnOssan_State_ContinueShoppingPrompt,
+        EnOssan_State_GiveLonLonMilk,
+        EnOssan_State_DisplayOnlyBombDialog,
+        EnOssan_State_WaitForDisplayOnlyBombDialog,
+        EnOssan_State_21,
+        EnOssan_State_22,
+        EnOssan_State_QuickBuyDialog,
+        EnOssan_State_SelectMaskItem,
+        EnOssan_State_LendMaskOfTruth,
+        EnOssan_State_GiveDiscountDialog,
+    };
+    Player* player = PLAYER;
+
+    this->blinkFunc(this);
+    EnOssan_UpdateJoystickInputState(globalCtx, this);
+    EnOssan_UpdateItemSelectedProperty(this);
+    EnOssan_UpdateStickDirectionPromptAnim(this);
+    EnOssan_UpdateCursorAnim(this);
+    Math_StepToS(&this->headRot, this->headTargetRot, 0x190);
+
+    if (player != NULL) {
+        stateFunc[this->stateFlag](this, globalCtx, player);
+    }
+
+    Actor_MoveForward(&this->actor);
+    Actor_UpdateBgCheckInfo(globalCtx, &this->actor, 26.0f, 10.0f, 0.0f, 5);
+    Actor_SetFocus(&this->actor, 90.0f);
+    Actor_SetScale(&this->actor, sShopkeeperScale[this->actor.params]);
+
+    // use animation object if needed
+    if (this->obj3ToSeg6Func != NULL) {
+        this->obj3ToSeg6Func(this, globalCtx);
+    }
+
+    SkelAnime_Update(&this->skelAnime);
+}
+
+void EnOssan_Update(Actor* thisx, GlobalContext* globalCtx) {
+    EnOssan* this = THIS;
+
+    this->timer++;
+    this->actionFunc(this, globalCtx);
+}
+
+s32 EnOssan_OverrideLimbDrawDefaultShopkeeper(GlobalContext* globalCtx, s32 limbIndex, Gfx** dList, Vec3f* pos,
+                                              Vec3s* rot, void* thisx) {
+    EnOssan* this = THIS;
+
+    if (limbIndex == 8) {
+        rot->x += this->headRot;
+    }
+    return 0;
+}
+
+void EnOssan_DrawCursor(GlobalContext* globalCtx, EnOssan* this, f32 x, f32 y, f32 z, u8 drawCursor) {
+    s32 ulx, uly, lrx, lry;
+    f32 w;
+    s32 dsdx;
+
+    OPEN_DISPS(globalCtx->state.gfxCtx, "../z_en_oB1.c", 4192);
+    if (drawCursor != 0) {
+        func_80094520(globalCtx->state.gfxCtx);
+        gDPSetPrimColor(OVERLAY_DISP++, 0, 0, this->cursorColor.r, this->cursorColor.g, this->cursorColor.b,
+                        this->cursorColor.a);
+        do {
```

Comment:
Is this actually needed? If it is maybe we should look at wrapping the macro it self with a do while.

Reply:
yes it's required